source/extensions/filters/http/transformation/inja_transformer.cc

```diff
@@ -65,6 +68,26 @@ Extractor::Extractor(const envoy::api::v2::filter::http::Extraction &extractor)
         fmt::format("group {} requested for regex with only {} sub groups",
                     group_, extract_regex_.mark_count()));
   }
+
+  // if replace_all is set, we must have replacement text
+  if (replace_all_ && !has_replacement_text_) {
+    throw EnvoyException(
+        fmt::format("replace_all set but no replacement text provided"));
+  }
+
+  // if replace_all is set, subgroup should be 0
+  if (replace_all_ && group_ != 0) {
+    throw EnvoyException(
+        fmt::format("replace_all set but subgroup is not 0"));
```

Comment:
Good catch, I should've highlighted this because I was unsure about this during implementation
 
Subgroup is really just not considered if we're in the `replace_all` mode. We definitely could just log a warning in that case.

The thing that influenced my decision to throw an exception is that the `regex` field has slightly different constraints depending on the mode selected (extract vs. single replace vs. replace all)

https://github.com/solo-io/envoy-gloo/blob/f747a124c15c56a09d6436607f4efc992e37859d/api/envoy/config/filter/http/transformation/v2/transformation_filter.proto#L178-L187

My thinking is that I don't want users to feel like they can toggle between modes on the fly -- if you have a working single replace extraction, it will not work the way you expect if you switch the mode to replace all. In that case, if you specify a subgroup (which I expect is typically the case), the system will indicate that the configuration is invalid, hopefully leading the user to realize that they may need more significant changes to their configuration.

Reply:
Aha, that makes a lot of sense. Thanks for the explanation. Given the anticipated use case of users hotswapping or copy/pasting config and changing the mode, I think it makes sense to enforce that. That said, I feel like that enforcement should occur on the control plane so that we are not throwing an exception on config. I see that we have precedent for doing that already, so I'm not super opinionated about it...

Def think we should enforce in the control plane though